source/request_response.c

```diff
@@ -831,6 +829,101 @@ struct aws_http_stream *aws_http_connection_make_request(
     return stream;
 }
 
+struct aws_http_message *aws_http2_message_new_from_http1(
+    struct aws_http_message *http1_msg,
+    struct aws_allocator *alloc) {
+
+    struct aws_http_headers *old_headers = aws_http_message_get_headers(http1_msg);
+    struct aws_http_header header_iter;
+    struct aws_byte_buf lower_name_buf;
+    AWS_ZERO_STRUCT(lower_name_buf);
+    struct aws_http_message *message = aws_http_message_is_request(http1_msg) ? aws_http2_message_new_request(alloc)
+                                                                              : aws_http2_message_new_response(alloc);
+    struct aws_http_headers *copied_headers = message->headers;
+    if (!message) {
+        return NULL;
+    }
+    /* Set pseudo headers from HTTP/1.1 message */
+    struct aws_byte_cursor method;
+    if (aws_http_message_get_request_method(http1_msg, &method)) {
+        AWS_LOGF_ERROR(
+            AWS_LS_HTTP_GENERAL,
+            "Failed to create http2 message from http1 message, ip: %p, due to no method found.",
+            (void *)http1_msg);
+        /* error will happen when the request is invalid */
+        aws_raise_error(AWS_ERROR_HTTP_INVALID_METHOD);
+        goto error;
+    }
+    /* Use add intead of set method to avoid push front to the array list */
+    if (aws_http_headers_add(copied_headers, aws_http_header_method, method)) {
+        goto error;
+    }
+    /**
+     * we set a default value, "https", for now.
+     * TODO: as we support prior knowledge, we may also want to support http?
+     */
+    struct aws_byte_cursor scheme_cursor = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("https");
+    if (aws_http_headers_add(copied_headers, aws_http_header_scheme, scheme_cursor)) {
+        goto error;
+    }
+
+    /* if host header found, add it as :authority */
+    struct aws_byte_cursor authority_cursor;
```

Comment:
Hmmm, a closer reading of the spec makes it seem like we should NOT create `authority` if the user didn't put it there themselves
https://httpwg.org/specs/rfc7540.html#rfc.section.8.1.2.3